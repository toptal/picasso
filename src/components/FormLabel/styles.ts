import { Theme, createStyles } from '@material-ui/core/styles'

import { alpha, rem } from '../styles'

export default ({ palette }: Theme) =>
  createStyles({
    root: {
      color: palette.grey[400]
    },

    disabled: {
      color: alpha(palette.grey[400], 0.48)
    },

    asterisk: {
      marginRight: rem('5px'),
      color: palette.error.main
    },

    block: {
      fontSize: '0.875em',
      display: 'block',
      marginBottom: rem('8px')
    },

    inline: {
      display: 'inline-block'
    }
  })
