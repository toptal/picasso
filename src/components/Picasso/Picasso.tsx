import {
  createMuiTheme,
  MuiThemeProvider,
  withStyles
} from '@material-ui/core/styles'
import React, { FunctionComponent, ReactNode } from 'react'

import CssBaseline from '../CssBaseline'
import {
  palette,
  layout,
  transitions,
  typography,
  spacing,
  breakpoints,
  screens,
  shadows
} from './config'
import FontsLoader from './FontsLoader'
import Provider from './PicassoProvider'
import globalStyles from './styles'
import { JssProps } from './types'

const picasso = {
  palette,
  layout,
  transitions,
  spacing,
  breakpoints,
  screens,
  shadows,
  typography,
  props: {
    MuiButtonBase: {
      disableRipple: true
    },
    MuiList: {
      disablePadding: true
    },
    MuiPaper: {
      square: true
    },
    MuiOutlinedInput: {
      notched: false
    }
  }
}

const PicassoProvider = new Provider(createMuiTheme(picasso))

interface PicassoGlobalStylesProviderProps extends JssProps {
  children?: ReactNode
}

const PicassoGlobalStylesProvider = withStyles(globalStyles, {
  name: 'Picasso'
})((props: PicassoGlobalStylesProviderProps) => {
  const { classes, children } = props

  return <div className={classes.root}>{children}</div>
})

interface PicassoProps {
  children?: ReactNode
  /** Whether to load fonts file to the page */
  loadFonts?: boolean
  /** Whether to apply Picasso CSS reset */
  reset?: boolean
}

const Picasso: FunctionComponent<PicassoProps> = ({
  loadFonts,
  reset,
  children
}) => (
  <MuiThemeProvider theme={PicassoProvider.theme}>
    {loadFonts && <FontsLoader />}
    {reset && <CssBaseline />}
    <PicassoGlobalStylesProvider>{children}</PicassoGlobalStylesProvider>
  </MuiThemeProvider>
)

Picasso.defaultProps = {
  loadFonts: true,
  reset: true
}

export { PicassoProvider }
export default Picasso
