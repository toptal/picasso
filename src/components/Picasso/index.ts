export {
  default,
  PicassoProvider,
  usePicassoRoot,
  usePageHeader
} from './Picasso'

export {
  BaseProps,
  JssProps,
  StandardProps,
  SizeType,
  OmitInternalProps,
  SpacingType,
  spacingToEm,
  PicassoComponent,
  PicassoComponentWithRef,
  SpacingEnum,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef,
  OverridableComponent,
  ColorType
} from './types'

export { useScreenSize, isScreenSize, useBreakpoint } from './config'
