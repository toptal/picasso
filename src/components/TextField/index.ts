import Input from '../Input'
import withDeprecationWarning from '../utils/with-deprecation-warning'

// TextField component is deprecated and
// have to be removed in v4 of Picasso
export default withDeprecationWarning('TextField', 'Input')(Input)
