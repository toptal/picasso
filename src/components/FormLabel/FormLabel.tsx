import React, { FunctionComponent } from 'react'
import { withStyles } from '@material-ui/core/styles'
import cx from 'classnames'

import { StandardProps } from '../Picasso'
import styles from './styles'

type ComponentType = 'label' | 'span'

export interface Props extends StandardProps {
  /** Content of the label */
  children: string
  /** Adds asteriks if true */
  required?: boolean
  /** Is this label for disabled input or not */
  disabled?: boolean
  /** Specifies an id of the input */
  htmlFor?: string
  /** Whether label should act as inline element `display: inline-block` */
  inline?: boolean
  /** Component used for the root node */
  as?: ComponentType
}

export const FormLabel: FunctionComponent<Props> = ({
  children,
  required,
  disabled,
  htmlFor,
  classes,
  className,
  style,
  inline,
  as: Component = 'label'
}) => {
  const isInline = inline || Component === 'span'

  return (
    <Component
      htmlFor={htmlFor}
      className={cx(
        classes.root,
        {
          [classes.disabled]: disabled,
          [classes.inline]: isInline
        },
        className
      )}
      style={style}
    >
      {required && <span className={classes.asterisk}>*</span>}
      <span className={classes.text}>{children}</span>
    </Component>
  )
}

FormLabel.defaultProps = {
  as: 'label',
  inline: false
}

FormLabel.displayName = 'FormLabel'

export default withStyles(styles)(FormLabel)
