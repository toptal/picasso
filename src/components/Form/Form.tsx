import React, { FunctionComponent, FormEventHandler, ReactNode } from 'react'

import FormField from '../FormField'
import FormHint from '../FormHint'
import FormLabel from '../FormLabel'
import FormError from '../FormError'
import { BaseProps } from '../Picasso'

interface Props extends BaseProps {
  /** Content of Form constructed of Form elements */
  children?: ReactNode
  /** Submit handler */
  onSubmit?: FormEventHandler<HTMLFormElement>
}

interface StaticProps {
  Field: typeof FormField
  Hint: typeof FormHint
  Label: typeof FormLabel
  Error: typeof FormError
}

export const Form: FunctionComponent<Props> & StaticProps = ({
  onSubmit,
  className,
  style,
  children
}) => (
  <form onSubmit={onSubmit} className={className} style={style}>
    {children}
  </form>
)

Form.Field = FormField
Form.Hint = FormHint
Form.Label = FormLabel
Form.Error = FormError

Form.displayName = 'Form'

export default Form
