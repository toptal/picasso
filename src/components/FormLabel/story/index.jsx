import PicassoBook from '~/.storybook/components/PicassoBook'

import { FormLabel } from '../FormLabel'

const chapter = PicassoBook.connectToPage(page =>
  page
    .createChapter(
      'Form Label',
      `Add the <Form.Label> component if you need to add a Label to your input`
    )
    .addExample('FormLabel/story/Default.example.jsx', 'Default')
    .addExample('FormLabel/story/Disabled.example.jsx', 'Disabled')
    .addExample('FormLabel/story/Required.example.jsx', 'Required')
)

const componentDocs = PicassoBook.createComponentDocs(FormLabel, 'Form.Label')

export default { chapter, componentDocs }
