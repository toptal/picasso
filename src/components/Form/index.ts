import { OmitInternalProps } from '../Picasso'
import { Props } from './Form'
export { default } from './Form'
export type Props = OmitInternalProps<Props>
