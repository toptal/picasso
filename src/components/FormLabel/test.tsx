import React, { FunctionComponent } from 'react'
import { render } from 'react-testing-library'
import 'react-testing-library/cleanup-after-each'

import FormLabel, { Props } from './FormLabel'
import Form from '../Form'
import TextField from '../TextField'
import Picasso, { OmitInternalProps } from '../Picasso'

const TestFormLabel: FunctionComponent<OmitInternalProps<Props>> = ({
  children,
  required,
  disabled,
  htmlFor,
  inline
}) => (
  <Picasso loadFonts={false}>
    <Form>
      <Form.Field>
        <FormLabel
          required={required}
          disabled={disabled}
          htmlFor={htmlFor}
          inline={inline}
        >
          {children}
        </FormLabel>
        <TextField />
      </Form.Field>
    </Form>
  </Picasso>
)

describe('FormLabel', () => {
  test('default render', () => {
    const { container } = render(<TestFormLabel>Label</TestFormLabel>)

    expect(container).toMatchSnapshot()
  })

  test('disabled', () => {
    const { container } = render(<TestFormLabel disabled>Label</TestFormLabel>)

    expect(container).toMatchSnapshot()
  })

  test('required', () => {
    const { container } = render(<TestFormLabel required>Label</TestFormLabel>)

    expect(container).toMatchSnapshot()
  })

  test('requried and disabled', () => {
    const { container } = render(
      <TestFormLabel required disabled>
        Label
      </TestFormLabel>
    )

    expect(container).toMatchSnapshot()
  })
})
