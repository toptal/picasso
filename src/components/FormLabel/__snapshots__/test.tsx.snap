// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLabel default render 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <form>
      <div
        class="FormField-root"
      >
        <label
          class="FormLabel-root"
        >
          <span
            class="FormLabel-text"
          >
            Label
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root OutlinedInput-root TextField-root OutlinedInput-rootAuto"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
          <input
            class="MuiInputBase-input MuiOutlinedInput-input OutlinedInput-input TextField-input"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`FormLabel disabled 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <form>
      <div
        class="FormField-root"
      >
        <label
          class="FormLabel-root FormLabel-disabled"
        >
          <span
            class="FormLabel-text"
          >
            Label
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root OutlinedInput-root TextField-root OutlinedInput-rootAuto"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
          <input
            class="MuiInputBase-input MuiOutlinedInput-input OutlinedInput-input TextField-input"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`FormLabel required 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <form>
      <div
        class="FormField-root"
      >
        <label
          class="FormLabel-root"
        >
          <span
            class="FormLabel-asterisk"
          >
            *
          </span>
          <span
            class="FormLabel-text"
          >
            Label
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root OutlinedInput-root TextField-root OutlinedInput-rootAuto"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
          <input
            class="MuiInputBase-input MuiOutlinedInput-input OutlinedInput-input TextField-input"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`FormLabel requried and disabled 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <form>
      <div
        class="FormField-root"
      >
        <label
          class="FormLabel-root FormLabel-disabled"
        >
          <span
            class="FormLabel-asterisk"
          >
            *
          </span>
          <span
            class="FormLabel-text"
          >
            Label
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root OutlinedInput-root TextField-root OutlinedInput-rootAuto"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend"
              style="width: 0.01px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
          <input
            class="MuiInputBase-input MuiOutlinedInput-input OutlinedInput-input TextField-input"
            type="text"
            value=""
          />
        </div>
      </div>
    </form>
  </div>
</div>
`;
